File: Dynamic_Programming/76.Longest Increasing Subsequence/Solution_DP.py
class Solution:
    """
    @param nums: An integer array
    @return: The length of LIS (longest increasing subsequence)
    """

    def longestIncreasingSubsequence(self, nums):
        # write your code here
        if not nums or len(nums) == 0:
            return 0

        # initialization
        # f[i] represents the max length of increasing subsequence which ends with nums[i]
        n = len(nums)
        f = [1 for _ in range(n)]

        # function: f[i] <- f[j] with contraints on j < i, nums[i] > nums[j]
        for i in range(1, n):
            for j in range(0, i):
                if nums[i] > nums[j]:
                    f[i] = max(f[i], f[j] + 1)

        return max(f)

File: Dynamic_Programming/76.Longest Increasing Subsequence/Solution_DP.java
public class Solution {
    /**
     * @param nums: An integer array
     * @return: The length of LIS (longest increasing subsequence)
     */
    public int longestIncreasingSubsequence(int[] nums) {
        // write your code here

        if (nums == null || nums.length == 0) {
            return 0;
        }

        int[] arr = new int[nums.length];
        Arrays.fill(arr, 1);

        int length = nums.length;
        for (int i = 0; i < length; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    arr[i] = Math.max(arr[j] + 1, arr[i]);
                }
            }
        }

        int size = 0;
        for (int idx = 0; idx < length; idx++) {
            if (arr[idx] > size) {
                size = arr[idx];
            }
        }

        return size;
    }

}